# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [7]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [11]:
# load as Pandas dataframe
df = pd.read_json('data/world_bank_projects.json')
df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

*Question 1*

First we need to find the column that carries the value for number of projects.


In [57]:
df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

We check for columns that have NANs in order to extract data later

In [58]:
df.isna().sum()[:5]

_id                      0
approvalfy               0
board_approval_month     0
boardapprovaldate        0
borrower                15
dtype: int64

We proceed to group by countryname, count ocurrences and extract the values for id which has no missing values

In [59]:
top10 = df.groupby(['countryname']).count()._id.sort_values(ascending=False)[:10]
top10

countryname
People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
Name: _id, dtype: int64

*Question 2*

The important column here is 'mjtheme_namecode'.
We proceed to extract the list as a series, transform it in to a list of lists of dictionaries, flatten it into a single list and build a new dataframe out of it

In [176]:
lst = df.mjtheme_namecode.tolist()
flatlist = [item for sublist in lst for item in sublist]
df2 = pd.DataFrame(flatlist)
df2.groupby('name').count().sort_values('code',ascending=False)[:10]

code
name                                              
Environment and natural resources management   223
Rural development                              202
Human development                              197
Public sector governance                       184
Social protection and risk management          158
Financial and private sector development       130
                                               122
Social dev/gender/inclusion                    119
Trade and integration                           72
Urban development                               47

In [177]:
df2.name.isin(['']).sum()

122

However this list is incomplete because several of the projects have no names; in fact there are 122 rows where the code is present yet the name is missing. 
We will proceed to fill the names according to their code and present an updated list of top 10 themes. First we check there are no missing codes:

In [178]:
print(df2.code.describe())
df2.code.isna().sum()

count     1499
unique      11
top         11
freq       250
Name: code, dtype: object


0

We see that the data type is object thuslets check if there are any non numeric values in the set:

In [179]:
print(pd.to_numeric(df2.code, errors='coerce').isna().sum())

0


With no errors in the code we change datatype to numeric:

In [180]:
df2.code = pd.to_numeric(df2.code, errors='coerce')

To proceed we need to know which codes go with which names:

In [181]:
df2.sort_values('code').groupby('code').max()

name
code                                              
1                              Economic management
2                         Public sector governance
3                                      Rule of law
4         Financial and private sector development
5                            Trade and integration
6            Social protection and risk management
7                      Social dev/gender/inclusion
8                                Human development
9                                Urban development
10                               Rural development
11    Environment and natural resources management

In [183]:
df2.loc[df2.code == 1, 'name'] = 'Economic management'
df2.loc[df2.code == 2, 'name'] = 'Public sector governance'
df2.loc[df2.code == 3, 'name'] = 'Rule of law'
df2.loc[df2.code == 4, 'name'] = 'Financial and private sector development'
df2.loc[df2.code == 5, 'name'] = 'Trade and integration'
df2.loc[df2.code == 6, 'name'] = 'Social protection and risk management'
df2.loc[df2.code == 7, 'name'] = 'Social dev/gender/inclusion'
df2.loc[df2.code == 8, 'name'] = 'Human development'
df2.loc[df2.code == 9, 'name'] = 'Urban development'
df2.loc[df2.code == 10, 'name'] = 'Rural development'
df2.loc[df2.code == 11, 'name'] = 'Environment and natural resources management'

In [188]:
df2.groupby('name').count().sort_values('code',ascending=False)[:10]

code
name                                              
Environment and natural resources management   250
Rural development                              216
Human development                              210
Public sector governance                       199
Social protection and risk management          168
Financial and private sector development       146
Social dev/gender/inclusion                    130
Trade and integration                           77
Urban development                               50
Economic management                             38

Now we have the complete list on the total number of theme for largest projects!